src/libmongoc/src/mongoc/mongoc-shared.c

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2021 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "./mongoc-shared.h"
+
+#include "common-thread-private.h"
+#include <bson/bson.h>
+
+#include <assert.h>
+
+typedef struct _mongoc_shared_ptr_aux {
+   int refcount;
+   void (*dtor) (void *);
+   void *managed;
+} _mongoc_shared_ptr_aux;
+
+static void
+_release_aux (_mongoc_shared_ptr_aux *aux)
+{
+   aux->dtor (aux->managed);
+   bson_free (aux);
+}
+
+static bson_mutex_t g_shared_ptr_spin_mtx = PTHREAD_MUTEX_INITIALIZER;
+
+static void
+_shared_ptr_spin_lock ()
+{
+   bson_mutex_lock (&g_shared_ptr_spin_mtx);
```

Comment:
I see this was changed from a spin lock in a prior commit. I have not thought about spin locks in a while, so please excuse the silly question.

IIRC spin locks loop and fully utilize the CPU. A mutex will make a system call into the kernel, and put the calling thread to sleep.

When the expectation is that the lock will not need to wait for very long, a spin lock is preferable to avoid the overhead of the system call of a mutex.

I am not sure which is preferable for shared pointer. Did the original spin lock run into problems?

Reply:
Yes, the original spin lock ran into problems. The issue is that it breaks the forward-progress guarantees of the lock-holder. If there are three threads on a 2-hw-thread system, and one thread is holding the spin lock and gets preempted, and then the other two threads attempt to take the spin lock, there is no indication to the kernel that they are spinning waiting for the sleeping thread to release the lock. If there are *hundreds* of threads blocking on the spinlock (as was happening in my benchmarking), it can almost appear to deadlock since all threads are waiting on the lock holder to get woken up by the kernel, which can take an arbitrary amount of time since the kernel doesn't know that it needs to do so. This can be fixed by yielding the thread when you spin, but is less than optimal because the scheduler doesn't know why you're doing that.

I believe the best tradeoff is to use a limited number of `try_lock()` spins, which will never actually syscall, followed by a `lock()` that might syscall and actually sleep the thread. This is the approach taken by LLVM's libc++.

Reply:
> If there are three threads on a 2-hw-thread system, and one thread is holding the spin lock and gets preempted, and then the other two threads attempt to take the spin lock, there is no indication to the kernel that they are spinning waiting for the sleeping thread to release the lock. If there are hundreds of threads blocking on the spinlock (as was happening in my benchmarking), it can almost appear to deadlock since all threads are waiting on the lock holder to get woken up by the kernel, which can take an arbitrary amount of time since the kernel doesn't know that it needs to do so

Interesting, thank you for the thorough explanation. That sounds like a priority inversion problem – the thread holding the spin lock should be the highest priority to schedule but the kernel is not aware.

> I believe the best tradeoff is to use a limited number of try_lock() spins, which will never actually syscall, followed by a lock() that might syscall and actually sleep the thread. This is the approach taken by LLVM's libc++.

That sounds close to the implementation proposed in https://github.com/mongodb/mongo-c-driver/pull/840/files#diff-7591f9551d8904ff97695bc90f2d6b819afe82b783c4e798b046d597acd3880fR57 for `_lock_64bit_atomic`. That might be interesting to explore if we need to fine tune this.